src/mumble/MainWindow.cpp

```diff
@@ -899,43 +899,172 @@ void MainWindow::openUrl(const QUrl &url) {
+		if (!namePrefix.isNull()) {
+			name.prepend(SEP);
+			name.prepend(namePrefix);
+		}
+
+		Channel *chan = searchChannelV1(parent, name);
+
+		return searchChannelV1(chan ? chan : parent, subPath, chan ? QString() : name, retExactMatch);
+	}
+
+	// Depth first search (not searching for alternative exact matches that are
+	// possible through ambiguous slashes or case)
+	static Channel *searchChannelV1(Channel *root, const QString &path, bool &retExactMatch) {
+		// The hostname-path separator slash is missing for URLs copied from:
+		// * The server (connect dialog)
+		// * The root channel
+		// * Channels with a leading slash that are a child of the root channel
+		//
+		// However, as path was passed through QUrl it always has a leading slash.
+		// It is thus ambiguous whether it is a channel name slash or separator.
+		//
+		// Most/All special characters are escaped, except slashes. Thus, a path
+		// may not be unique (c1/c2 could be a channel or parent and child).
+		//
+		// Channels with a leading slash deeper into the tree have their own
+		// slash (path separator + channel name slash: root/chan1//ChanwithOneLeadingSlash).
+
+		QStringList qlChans = path.split(SEP);
+		Channel *chan = searchChannelV1(root, qlChans, QString(), retExactMatch);
+
+		// We expect a path of "/" to have been an URL without a path slash
+		// in most cases (an URL to the server/root channel). In that case, the
+		// root channel is an exact match. As "/" is a valid channel as well, we
+		// had to look for it, but did not find such a channel.
+		if (!retExactMatch && path == SEP) {
+
+			retExactMatch = true;
+		}
+		return chan;
+	}
+
+	static Channel *searchChannelV2(Channel *parent, QStringList &chans, bool &retExactMatch) {
+		if (chans.isEmpty()) {
+			retExactMatch = true;
+			return parent;
+		}
+
+		QString name = chans.takeFirst();
+		foreach(Channel *c, parent->qlChannels) {
+			QString cn = getEncodedChannelname(c);
+			if (cn == name) {
+				return searchChannelV2(c, chans, retExactMatch);
+			}
+		}
+		retExactMatch = false;
+		return parent;
+	}
+
+	static Channel *searchChannelV2(Channel *root, const QString &path, bool &retExactMatch) {
+		QStringList qlChans = path.split(SEP);
+		// The path always begins with a slash - remove the first element
+		qlChans.removeFirst();
+		return searchChannelV2(root, qlChans, retExactMatch);
+	}
+
+public:
+	static QString getUrlPathForChannel(Channel *c) {
+		// For root channel, use (only) a slash
+		if (c->cParent == NULL) {
+			return SEP;
+		}
+		QString path;
+		while (c->cParent != NULL) {
+			path.prepend(getEncodedChannelname(c));
+			path.prepend(SEP);
+			c = c->cParent;
+		}
+		return path.prepend(SEP);
+	}
+
+	static Channel *searchChannelByPath(QString &path, bool &retExactMatch) {
+		Channel *root = Channel::get(0);
+		if (!root) {
+			g.l->log(Log::CriticalError, QString::fromUtf8("Failed to identify Root channel"));
```

Comment:
I wouldn’t actually expect this to ever happen. I kept it in as it was being checked before, but would like to suggest removing it. I don’t know the code though, but I would expect us to always have a root channel.

If we do not always have a root channel, opening a desired channel should not be a Fatal though. If we join the server fine, that’s already half the way. Of course you could argue this function does not know the context it is being used in - but then it could just return NULL as well.